src/utilities/timeline.ts

```diff
@@ -906,97 +941,212 @@ export function spanInView(span: Span, viewTimeRange: TimeRange) {
+              if (childSpan && hierarchyMode === 'flat') {
+                seenSpans[childSpan.span_id] = true;
+              }
             }
-          }
-          if (expanded) {
-            children.push(
-              getDirectiveSubtree(
-                directive,
-                id,
-                activityTreeExpansionMap,
-                filterActivitiesByTime,
-                spanUtilityMaps,
-                spansMap,
-                showSpans,
-                viewTimeRange,
-              ),
-            );
-          }
-          items.push({ directive, ...(childSpan ? { span: childSpan } : null) });
-        });
-      }
-      if (spanGroup && hierarchyMode === 'flat') {
-        spanGroup.forEach(span => {
-          if (!seenSpans[span.span_id]) {
             if (expanded) {
               children.push(
-                ...getSpanSubtrees(
-                  span,
+                getDirectiveSubtree(
+                  directive,
                   id,
-                  activityTreeExpansionMap,
-                  'span',
+                  discreteTreeExpansionMap,
                   filterActivitiesByTime,
                   spanUtilityMaps,
                   spansMap,
+                  showSpans,
                   viewTimeRange,
                 ),
               );
             }
-            items.push({ span });
-          }
+            items.push({ directive, ...(childSpan ? { span: childSpan } : null) });
+          });
+        }
+        if (spanGroup && hierarchyMode === 'flat') {
+          spanGroup.forEach(span => {
+            if (!seenSpans[span.span_id]) {
+              if (expanded) {
+                children.push(
+                  ...getSpanSubtrees(
+                    span,
+                    id,
+                    discreteTreeExpansionMap,
+                    'span',
+                    filterActivitiesByTime,
+                    spanUtilityMaps,
+                    spansMap,
+                    viewTimeRange,
+                  ),
+                );
+              }
+              items.push({ span });
+            }
+          });
+        }
+        activityNodes.push({
+          activity_type: 'aggregation',
+          children: paginateNodes(children, 'activity', id, discreteTreeExpansionMap),
+          expanded: expanded,
+          id,
+          isLeaf: false,
+          items,
+          label,
+          type: 'Activity',
         });
-      }
-      nodes.push({
-        children: paginateNodes(children, id, activityTreeExpansionMap),
-        expanded: expanded,
-        id,
-        isLeaf: false,
-        items,
-        label,
-        type: 'aggregation',
       });
-    });
-  return nodes;
+  }
+
+  // make the external event subtree
+  const externalEventNodes: DiscreteTreeNode[] = [];
+  if (hasExternalEventsLayer) {
+    if (Object.keys(groupedExternalEvents).length !== 0) {
+      const allKeys = new Set(Object.keys(groupedExternalEvents));
```

Comment:
I don't think you need to convert to a Set here. If you're using it to get an array of unique keys, an object will inherently already have a unique set of keys.